### PR TITLE
Fix singular and plural search results text

### PR DESCRIPTION
**Pull request by hugovk**
### Feature or Bugfix

- Bugfix

### Purpose
- Show accurate search results text.
 
### Detail

Search for something and, if there are results, it will say something like "Search finished, found 1 page(s) matching the search query." or "Search finished, found 23 page(s) matching the search query."

Computers are bad at some things, but one thing they're very good at is counting :)

Let's show "1 page" for a single result, and "${resultCount} pages" for multiple.

#### Before

Shows "page(s)" for both single and multiple results:

image

https://docs.python.org/3/search.html?q=hotspot

image

https://docs.python.org/3/search.html?q=lru_cache

#### After

Shows "page" for a single result and "pages" for multiple results:

image

image

### Localisation

Do we need to do something to update the translation files?

### CODE REVIEW

**Comment by jayaddison** [comment]
Note that there's an [`ngettext` interface](https://github.com/sphinx-doc/sphinx/blob/0d912c85fd3ec385432fe707f6a0678425d1e841/sphinx/themes/basic/static/doctools.js#L58-L63) available in the JS code, intended to allow count-aware internationalization of messaging.

https://github.com/sphinx-doc/sphinx/blob/0d912c85fd3ec385432fe707f6a0678425d1e841/sphinx/locale/sr/LC_MESSAGES/sphinx.js#L59

Could we use that to achieve a more internationalizable fix?

**Comment by hugovk** [comment]
That sounds better, do you know how to use it?

**Comment by jayaddison** [comment]
> That sounds better, do you know how to use it?

I don't yet, no, but I'd be willing to learn in support of reviewing and/or suggesting an adjustment of the PR that uses it.  That could take me a few days.

**Comment by jayaddison** [comment]
Ah, and I remembered that I should probably spend some time refreshing my knowledge of other features to support an upcoming 8.0.0 release - so I'd update my timeline here to a couple of weeks or so to provide support (possibly pessimistic, but I'd prefer that than overcommitting).

**Comment by hugovk** [comment]
Yeah, no rush at all for this one :)

---

Right, so `_` is an alias for `Documentation.gettext`, and there's an `ngettext` for dealing with singular and plural forms:

https://www.gnu.org/software/gettext/manual/html_node/Plural-forms.html

The next step is updating the translation files with these strings.

**Comment by hugovk** [comment]
There are some instructions at https://www.sphinx-doc.org/en/master/internals/contributing.html#translations

I ran `python babel_runner.py extract` and `python babel_runner.py update` but not `python babel_runner.py compile`.

**Comment by jayaddison** [comment]
That's great - I wasn't certain whether `babel_runner.py` would detect the `ngettext(...)` call - I thought we might have to adjust the callsite to use `__(...)` instead, based on the [defined keywords](https://github.com/sphinx-doc/sphinx/blob/0d912c85fd3ec385432fe707f6a0678425d1e841/utils/babel_runner.py#L76).  But indeed [`ngettext` is in the `babel` defaults](https://github.com/python-babel/babel/blob/76677eacd10cc35bea0a11ec1d2f0566cadd4066/babel/messages/extract.py#L79-L90) (and must pattern-match OK regardless of reference from a `Documentation` class).

File: sphinx/themes/basic/static/searchtools.js [hunk 1]
```diff
@@ -111,12 +111,14 @@ const _finishSearch = (resultCount) => {
   Search.stopPulse();
   Search.title.innerText = _("Search Results");
   if (!resultCount)
-    Search.status.innerText = Documentation.gettext(
+    Search.status.innerText = _(
```
Comment (jayaddison) [inline sphinx/themes/basic/static/searchtools.js]: Is this change necessary?

It's nitpicky, but I'd generally both to avoid modifying unnecessary code lines, and also to maintain consistency (`Documentation.gettext` alongside `Document.ngettext`, compared to usage of underscore abbreviation in one but not the other.  I admit that the abbreviated forms are often useful).

File: sphinx/locale/tr/LC_MESSAGES/sphinx.po [hunk 2]
```diff
@@ -3654,24 +3874,27 @@ msgstr "Arama Sonuçları"
...
+"yazdığınızdan ve gerekli bütün kategorileri seçtiğinizden emin olun."
 
 #: sphinx/themes/basic/static/searchtools.js:118
-msgid ""
-"Search finished, found ${resultCount} page(s) matching the search query."
-msgstr ""
+msgid "Search finished, found one page matching the search query."
+msgid_plural "Search finished, found ${resultCount} pages matching the search query."
+msgstr[0] ""
+msgstr[1] ""
```
Comment (jayaddison) [inline sphinx/locale/tr/LC_MESSAGES/sphinx.po]: Ok, that's a good confirmation; previously `msgid_plural` hadn't appeared in any resource catalogs :+1:

File: sphinx/themes/basic/static/searchtools.js (hunk 1 above)
Comment (hugovk) [reply sphinx/themes/basic/static/searchtools.js]: It's not necessary, happy to revert if you like, or make a separate PR with just this change?

I initially changed it because the `_` alias is used everywhere else in this file. That was before adding `Documentation.ngettext`, so I can see the case for symmetry with `Documentation.gettext`.

Personally I would use `_` where possible, so there aren't three forms floating around (`_`, `Documentation.gettext`, `Documentation.ngettext`) but no strong opinion.

Comment (jayaddison) [reply sphinx/themes/basic/static/searchtools.js]: It does probably seem pedantic, but yep, I think I'd prefer undoing this single-line change.

And then as a follow-up, I'd suggest adding a quick-alias in `doctools.js` for `__` (to complement the existing `_` alias), and migrating both callsites to use that.

Comment (jayaddison) [reply sphinx/themes/basic/static/searchtools.js]: (the latter I suppose should be considered a feature, despite being a small one)

**Comment by jayaddison** [comment]
This will also need a `CHANGES.rst` entry - those usually reference a separate bugreport number, but it is also valid to self-reference the PR number.

**Review by jayaddison — APPROVED** [review]
Looks good to me :+1: 

This is a nice improvement for the quality of search result presentation and localization.

**Review by firefly9966 — APPROVED** [review]

**Comment by jayaddison** [comment]
@AA-Turner aside from the merge conflict resolution, is there anything else to resolve before this could be considered for merge?

**Review by AA-Turner — APPROVED** [review]

**Comment by jayaddison** [comment]
Thank you @hugovk @AA-Turner!
